packages/assets-controllers/src/balances.ts

```diff
@@ -0,0 +1,506 @@
+    )[walletId];
+    if (!wallet) {
+      continue;
+    }
+
+    const groups = Object.keys(wallet.groups || {}) as string[];
+    for (const groupId of groups) {
+      const accounts = getInternalAccountsForGroup(
+        accountTreeState,
+        accountsState,
+        groupId,
+      );
+
+      for (const account of accounts) {
+        const isEvmAccount = isEvmAccountType(account.type);
+
+        if (isEvmAccount) {
+          const accountBalances =
+            tokenBalancesState.tokenBalances[account.address as Hex];
+          if (!accountBalances) {
+            continue;
+          }
+
+          for (const [chainId, chainBalances] of Object.entries(
+            accountBalances,
+          )) {
+            if (!isEvmChainEnabled(chainId as Hex)) {
+              continue;
+            }
+            const chainMarketData = tokenRatesState.marketData[chainId as Hex];
+
+            for (const [tokenAddress, balance] of Object.entries(
+              chainBalances,
+            )) {
+              const chainTokens = tokensState.allTokens[chainId as Hex];
+              const accountTokens = chainTokens?.[account.address];
+              const token = accountTokens?.find(
+                (t) => t.address === tokenAddress,
+              );
+              if (!token) {
+                continue;
+              }
+
+              const decimals =
+                typeof token.decimals === 'number' &&
+                !Number.isNaN(token.decimals)
+                  ? token.decimals
+                  : 18;
+
+              const balanceInSmallestUnit = parseInt(balance as string, 16);
+              if (Number.isNaN(balanceInSmallestUnit)) {
+                continue;
+              }
+
+              const balanceInTokenUnits =
+                balanceInSmallestUnit / Math.pow(10, decimals);
+
+              const tokenMarketData = chainMarketData?.[tokenAddress as Hex];
+              const price = tokenMarketData?.price as number | undefined;
+              const percentRaw = (
+                tokenMarketData as unknown as Record<string, unknown>
+              )?.[evmPercentField[period] as string] as number | undefined;
+
+              if (typeof price !== 'number' || Number.isNaN(price)) {
+                continue;
+              }
+
+              const nativeCurrency = (
+                tokenMarketData as unknown as { currency?: string }
+              )?.currency;
+              const nativeToUserRate =
+                nativeCurrency &&
+                currencyRateState.currencyRates[nativeCurrency]
+                  ? currencyRateState.currencyRates[nativeCurrency]
+                      ?.conversionRate
+                  : undefined;
+
+              if (
+                typeof nativeToUserRate !== 'number' ||
+                Number.isNaN(nativeToUserRate)
+              ) {
+                continue;
+              }
+
+              if (typeof percentRaw !== 'number' || Number.isNaN(percentRaw)) {
```

Comment:
This type and `isNan` check is used in a handful of places in this file; could this be a helper function?